### PR TITLE
vmConfig check for invalid tracer.

### PR DESCRIPTION
**Pull request by 0xTylerHolmes**
When we create a new blockchain we supply a vmConfig object. If we set debug mode to true without supplying aTracer we have the following nil pointer deref.

https://github.com/ethereum/go-ethereum/blob/7ca4f60a1abd2a8c9df2ed9ae35a891b0ac9e464/core/state_transition.go#L328

If we specify debug without adding a tracer we will get the nil pointer deref here. To mitigate this the constructor for blockchain checks that if the vmConfig has debug mode enabled it also checks to make sure the tracer is non-nil.

### CODE REVIEW

**Comment by rjl493456442** [comment]
I think we can do it in an alternative approach. We can remove the Debug field in `vm.Config`, but define `Tracing()` function instead. If the tracer is configured, then we dump the execution statistics into tracer.

```go
// Config are the configuration options for the Interpreter
type Config struct {
	// Debug                   bool      // Enables debugging
	Tracer                  EVMLogger // Opcode logger
	NoBaseFee               bool      // Forces the EIP-1559 baseFee to 0 (needed for 0 price calls)
	EnablePreimageRecording bool      // Enables recording of SHA3/keccak preimages
	ExtraEips               []int     // Additional EIPS that are to be enabled
}

// Tracing returns an indicator if evm execution tracer is configured.
func (config *Config) Tracing() bool {
	return config.Tracer != nil
}
```

**Review by holiman — APPROVED** [review]
I was thinking we could just disable `Debug` and print a warning, but OTOH this is a situation where you probably would rather abort and retry than continue and see a warning. 
I'm fine with this. The solution from @rjl493456442 is "not great" IMO, because the `Debug` check happens in a lot ot hot-spots, where we don't want to incur a method call. 

If we switched to use proper constructor-invocations, then it would be easier to solve it by having the Debug flag become set implicitly rather than explicitly. 

TLDR; SGTM for now

**Comment by s1na** [comment]
I'm slightly in favor of dropping `Debug` as Gary suggested. We can check `vmConfig.Tracer != nil` to avoid the method invocation overhead.

It seems so far tracer is the only use-case for `Debug`. Even if it were other cases (e.g. mocking caller as per https://github.com/ethereum/go-ethereum/pull/26414) they won't be necessarily invoked in the same place as tracing. E.g. see:

https://github.com/ethereum/go-ethereum/pull/26414/files#diff-64508d317d86e7a4a5294d76455a5e74148e26883da9e8b4ffc9bbfa3cc8550eR289-R290

**Comment by holiman** [comment]
> I think we can do it in an alternative approach.

Implemented in #27048, PTAL
